### ISSUE TITLE
WebAssembly: module closed with exit_code(1)

### ISSUE THREAD

**Issue by kvdbergh** [klippa-app/go-pdfium#88]
For some documents combined with a specific page width the `RenderPageInPixels` function results in a `module closed with exit_code(1)` error.

I have tried setting `WithDebugInfoEnabled(true)` on the `wazero.NewRuntimeConfig()` with a debug build of pdfium.wasm, but I am unable to get any specific error details.

Attached is a sample document in which the error can be reproduced.

Width: 2600px error
Width: 2900px works

Test code:

```golang
package main

import (
	"github.com/klippa-app/go-pdfium/requests"
	"github.com/klippa-app/go-pdfium/webassembly"
	"math"
	"os"
	"time"
)

func main() {
	pool, err := webassembly.Init(webassembly.Config{
		MinIdle:  1,
		MaxIdle:  1,
		MaxTotal: 1,
	})

	pdfBytes, err := os.ReadFile("test.pdf")
	checkError(err)

	instance, err := pool.GetInstance(time.Second * 5)
	checkError(err)
	defer instance.Close()

	doc, err := instance.OpenDocument(&requests.OpenDocument{
		File: &pdfBytes,
	})
	checkError(err)

	defer instance.FPDF_CloseDocument(&requests.FPDF_CloseDocument{
		Document: doc.Document,
	})

	pageRequest := requests.Page{
		ByIndex: &requests.PageByIndex{
			Document: doc.Document,
			Index:    0,
		},
	}

	pageWidth, err := instance.FPDF_GetPageWidth(&requests.FPDF_GetPageWidth{Page: pageRequest})
	checkError(err)

	pageHeight, err := instance.FPDF_GetPageHeight(&requests.FPDF_GetPageHeight{Page: pageRequest})
	checkError(err)

	width := 2600
	ratio := float64(width) / pageWidth.Width
	height := int(math.Floor(pageHeight.Height * ratio))

	pageRender, err := instance.RenderPageInPixels(&requests.RenderPageInPixels{
		Page:        pageRequest,
		Width:       width,
		Height:      height,
		RenderFlags: 0,
	})
	checkError(err)

	pageRender.Cleanup()
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
```

[results.txt](https://github.com/klippa-app/go-pdfium/files/12249651/results.txt)
[test.pdf](https://github.com/klippa-app/go-pdfium/files/12249656/test.pdf)

**Comment by jerbob92** [comment]
`WithDebugInfoEnabled` is already true by default, it only helps if the crash is in Go side, not when the program just exits. I have traced the error back to the following:

```
gray_render_line
  gray_set_cell
    emscripten_longjmp
      setThrew
        _emscripten_throw_longjmp
          abort
            _Exit
              wasi_snapshot_preview1.proc_exit
                module closed with exit_code(1)
```

Basically what happened is that the code called something that Emscripten does not support right now in standalone mode. I don't know exactly what triggers it yet, I'll look into a fix.

**Comment by jerbob92** [comment]
So I have changed my patch a bit to not exit on `_emscripten_throw_longjmp`
Then the Wazero stacktrace is as following:

```
wasm error: unreachable
wasm stack trace:
	.gray_set_cell(i32,i32,i32)
	.gray_render_line(i32,i32,i32)
	.gray_render_conic(i32,i32,i32)
	.gray_conic_to(i32,i32,i32) i32
	.FT_Outline_Decompose(i32,i32,i32) i32 (recovered by wazero)
wasm stack trace:
	env.invoke_iiii(i32,i32,i32,i32) i32
	.gray_convert_glyph_inner(i32,i32) i32
	.gray_convert_glyph(i32) i32
	.gray_raster_render(i32,i32) i32
	.ft_smooth_raster_lcd(i32,i32,i32) i32
	.ft_smooth_render(i32,i32,i32,i32) i32
	.FT_Render_Glyph_Internal(i32,i32,i32) i32
	.FT_Render_Glyph(i32,i32) i32
	.CFX_GlyphCache::RenderGlyph(CFX_Font const*, unsigned int, bool, CFX_Matrix const&, int, int)(i32,i32,i32,i32,i32,i32,i32) i32
	.CFX_GlyphCache::LookUpGlyphBitmap(CFX_Font const*, CFX_Matrix const&, fxcrt::ByteString const&, unsigned int, bool, int, int)(i32,i32,i32,i32,i32,i32,i32,i32) i32
	.CFX_GlyphCache::LoadGlyphBitmap(CFX_Font const*, unsigned int, bool, CFX_Matrix const&, int, int, CFX_TextRenderOptions*)(i32,i32,i32,i32,i32,i32,i32,i32) i32
	.CFX_Font::LoadGlyphBitmap(unsigned int, bool, CFX_Matrix const&, int, int, CFX_TextRenderOptions*) const(i32,i32,i32,i32,i32,i32,i32) i32
	.CFX_RenderDevice::DrawNormalText(pdfium::span<TextCharPos const>, CFX_Font*, float, CFX_Matrix const&, unsigned int, CFX_TextRenderOptions const&)(i32,i32,i32,f32,i32,i32,i32) i32
	.CPDF_TextRenderer::DrawNormalText(CFX_RenderDevice*, pdfium::span<unsigned int const>, pdfium::span<float const>, CPDF_Font*, float, CFX_Matrix const&, unsigned int, CPDF_RenderOptions const&)(i32,i32,i32,i32,f32,i32,i32,i32) i32
	.CPDF_RenderStatus::ProcessText(CPDF_TextObject*, CFX_Matrix const&, CFX_Path*)(i32,i32,i32,i32) i32
	.CPDF_RenderStatus::ProcessObjectNoClip(CPDF_PageObject*, CFX_Matrix const&)(i32,i32,i32)
	.CPDF_RenderStatus::ContinueSingleObject(CPDF_PageObject*, CFX_Matrix const&, PauseIndicatorIface*)(i32,i32,i32,i32) i32
	.CPDF_ProgressiveRenderer::Continue(PauseIndicatorIface*)(i32,i32)
	.CPDF_ProgressiveRenderer::Start(PauseIndicatorIface*)(i32,i32)
	.(anonymous namespace)::RenderPageImpl(CPDF_PageRenderContext*, CPDF_Page*, CFX_Matrix const&, FX_RECT const&, int, FPDF_COLORSCHEME_ const*, bool, CPDFSDK_PauseAdapter*)(i32,i32,i32,i32,i32,i32,i32,i32)
	.CPDFSDK_RenderPageWithContext(CPDF_PageRenderContext*, CPDF_Page*, int, int, int, int, int, int, FPDF_COLORSCHEME_ const*, bool, CPDFSDK_PauseAdapter*)(i32,i32,i32,i32,i32,i32,i32,i32,i32,i32,i32)
	.FPDF_RenderPageBitmap(i32,i32,i32,i32,i32,i32,i32,i32)
```

So it errors somewhere in freetype and it looks like freetype is doing some invalid jump. Will have to investigate more .

**Comment by kvdbergh** [comment]
@jerbob92 Thanks for your work on this project and research on this issue.

**Comment by jerbob92** [comment]
@kvdbergh I have found the source of the issue, we have missed a crucial part of the setjmp/longjmp handling in Wazero so that's why it's not working, currently working with Wazero team to fix it!

**Comment by jerbob92** [comment]
@kvdbergh I have created a fix for Wazero: https://github.com/tetratelabs/wazero/pull/1611
I will have to rebuild pdfium to make a complete fix, will do once that has been merged.

**Comment by jerbob92** [comment]
This has been fixed in v1.7.0!

**Comment by kvdbergh** [comment]
Thanks for the fix!
